src/System.Text.Json/src/System/Text/Json/Serialization/Converters/DefaultIEnumerableConstructibleConveter.cs

```diff
@@ -0,0 +1,43 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Reflection;
+using System.Text.Json.Serialization.Policies;
+
+namespace System.Text.Json.Serialization.Converters
+{
+    internal sealed class DefaultIEnumerableConstructibleConverter : JsonEnumerableConverter
+    {
+        // This method is invoked with reflection in CreateFromList.
+        private static IEnumerable CreateFromListInternal<TEnumerable, TElement>(IList sourceList)
+        {
+            List<TElement> list = new List<TElement>();
+
+            foreach (object item in sourceList)
+            {
+                if (item is TElement itemTElement)
```

Comment:
Is being tolerant here the right behavior? 

(thinking about `item is TElement` vs `(TElement)item`). I usually use the c-style cast when I don't expect it to ever fail - then it will make noise instead of omitting something.

Reply:
@JamesNK when there is JSON that can't be applied I agree we should throw during deserialization. However what is your thinking when we don't know how to serialize a given type -- throw by default or ignore by default?

Reply:
What do you mean by not know how to serialize? Do you mean like deserializing to an `object`? In Json.NET if I have an object then I take whatever JSON matches it and read it into a JToken, i.e. its raw JSON value, and set that.

```cs
public class Person
{
    public object Hobbies { get; set; }
}
```

```json
{
    "Hobbies": [".NET", "May-mays"]
}
```

It's not perfect, some people expect a serializer to be magic and roundtrip everything, but IMO it is better than `null`.

Reply:
No just serialization. For example, if we don't support serializing IDictionary (non-generic version) or there is no "converter" for a given value type.

Reply:
Automatically ignoring them wouldn't be intuitive to the user. A user would be confused why one type is serialized to JSON and another is missing.

You should either throw an exception - I think this is what WCF serializers do - or support serializing them - which is what Json.NET does.

Reply:
Ideally yes we would support things like non-generic IDictionary etc. and not ignore or throw.

However I was curious as to whether Json.NET has cases where it throws during serialization. Currently we know we will throw if there is a cycle, but had discussions on whether we should throw for other cases.

For example, if I have a plain struct with no converter, what should it do:
- Serialize as an object (along with any public properties)
- Serialize as a value. If so, then we have to pick a default type (number, string)
- Throw
- Ignore

I think Throwing would be the most correct. A [JsonIgnore] would need to be added to the property to prevent the exception. However that is inconvenient or impossible if you don't own the type. So some sort of option perhaps.

We can take this offline...

Reply:
Json.NET will serialize a struct as if it were a class.

```
public struct Person
{
    public string Name { get; set; }
}

{ "Name": "Mr Struct" }
```

Reply:
Json.NET will never throw because it doesn't understand how to serialize something. Typically it will only throw because the underlying instance threw an error, e.g. serializing a property that throws a `NotImplementedException`, or the writer can't write to the destination, e.g. the network stream is closed.

Reply:
Hmm if we treat the unknown case as serialize to class\object but later want to support `IConvertible` (for example, we probably should) then we need to do the right thing in v1 (we can't change behavior later).

So IMO throwing `NotSupportedException` may be the best short-term tactic in these cases if we think we may change the semantics later.